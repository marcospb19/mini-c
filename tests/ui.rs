//! This is a test module that creates snapshots, and might do
//! UI testing in the future.
//!
//! This shouldn't be called ui.rs hmmmmmmmm.
//!
//! Snapshot todo list:
//! - [X] Source code.
//! - [X] AST.
//! - [ ] LLVM IR.
//! - [ ] Execution stdout output (UI).
//! - [ ] (maybe) Execution final state (variables, functions).

use fs_err as fs;
use mini_c::test_utils;

macro_rules! snapshot_example_file {
    ($example_path:literal) => {
        let file_contents = &fs::read_to_string($example_path)
            .unwrap_or_else(|err| panic!("Failed to read file {}: {err}.", $example_path));
        // Source code
        insta::assert_display_snapshot!(file_contents);
        // AST
        insta::assert_ron_snapshot!(test_utils::generate_ast(file_contents));
    };
}

#[test]
fn example_break() {
    snapshot_example_file!("examples/break.c");
}
#[test]
fn example_coins() {
    snapshot_example_file!("examples/coins.c");
}
#[test]
fn example_conds() {
    snapshot_example_file!("examples/conds.c");
}
#[test]
fn example_easy() {
    snapshot_example_file!("examples/easy.c");
}
#[test]
fn example_fib() {
    snapshot_example_file!("examples/fib.c");
}
#[test]
fn example_file_1() {
    snapshot_example_file!("examples/file_1.c");
}
#[test]
fn example_file_2() {
    snapshot_example_file!("examples/file_2.c");
}
#[test]
fn example_if_pass() {
    snapshot_example_file!("examples/if-pass.c");
}
#[test]
fn example_operators() {
    snapshot_example_file!("examples/operators.c");
}
#[test]
fn example_queen() {
    snapshot_example_file!("examples/queen.c");
}
#[test]
fn example_var() {
    snapshot_example_file!("examples/var.c");
}
#[test]
fn example_var_loop() {
    snapshot_example_file!("examples/var_loop.c");
}
#[test]
fn example_while() {
    snapshot_example_file!("examples/while.c");
}
