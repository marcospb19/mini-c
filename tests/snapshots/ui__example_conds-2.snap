---
source: tests/ui.rs
expression: test_mini_c(file_contents)
---
Program(
  preamble: true,
  declarations: [
    Function(FunctionDeclaration(Bool, "isTen", [
      Parameter(Int, "a"),
    ], Some(Scope([], [
      If(IfStatement(Binary(BinaryExpression(VariableReference(Normal("a")), NotEquals, Value(Int(10)))), Scope([], [
        Return(ReturnStatement(Some(Value(Bool(false))))),
      ]), None)),
      Return(ReturnStatement(Some(Value(Bool(true))))),
    ])))),
    Function(FunctionDeclaration(Int, "main", [], Some(Scope([
      VariableDeclaration(Int, [
        Normal("myScore"),
        Normal("yourScore"),
      ]),
    ], [
      Expression(Assignment(AssignmentExpression(Normal("myScore"), Value(Int(10))))),
      Expression(Assignment(AssignmentExpression(Normal("yourScore"), Unary(UnaryExpression(Negative, Value(Int(10))))))),
      If(IfStatement(Binary(BinaryExpression(FunctionCall(FunctionCallExpression("isTen", [
        VariableReference(Normal("myScore")),
      ])), And, FunctionCall(FunctionCallExpression("isTen", [
        VariableReference(Normal("yourScore")),
      ])))), Scope([], [
        Expression(FunctionCall(FunctionCallExpression("putint", [
          Value(Int(1)),
        ]))),
        Return(ReturnStatement(Some(Value(Int(1))))),
      ]), None)),
      Expression(FunctionCall(FunctionCallExpression("putint", [
        Value(Int(0)),
      ]))),
      Return(ReturnStatement(Some(Value(Int(0))))),
    ])))),
  ],
)
